lib/hsa/mcwamp_hsa.cpp

```diff
@@ -5288,29 +5298,32 @@ HSACopy::enqueueAsyncCopyCommand(const Kalmar::HSADevice *copyDevice, const hc::
 
 
         int depSignalCnt = 0;
-        hsa_signal_t depSignal;
+        hsa_signal_t depSignal = { .handle = 0x0 };
         setCommandKind (resolveMemcpyDirection(srcPtrInfo._isInDeviceMem, dstPtrInfo._isInDeviceMem));
 
         if (!hsaQueue()->nextSyncNeedsSysRelease()) {
             DBOUT( DB_CMD2, "  copy launching without adding system release\n");
         }
 
         auto fenceScope = (hsaQueue()->nextSyncNeedsSysRelease()) ? hc::system_scope : hc::no_scope;
-        depAsyncOp = std::static_pointer_cast<HSAOp> (hsaQueue()->detectStreamDeps(this->getCommandKind(), this));
 
-        // We need to ensure the copy waits for preceding commands the HCC queue to complete, if those commands exist.
-        // The copy has to be set so that it depends on the completion_signal of the youngest command in the queue.
+        depAsyncOp = std::static_pointer_cast<HSAOp> (hsaQueue()->detectStreamDeps(this->getCommandKind(), this));
         if (depAsyncOp) {
             depSignal = * (static_cast <hsa_signal_t*> (depAsyncOp->getNativeHandle()));
+        }
 
+        // We need to ensure the copy waits for preceding commands the HCC queue to complete, if those commands exist.
+        // The copy has to be set so that it depends on the completion_signal of the youngest command in the queue.
+        if (depAsyncOp || fenceScope != hc::no_scope) {
```

Comment:
I believe this corresponds to the 2nd item in the description:

```
before a copy, even when there's no dependent async op, if the queue indicates that the next sync needs a system release, enqueue a marker with system scope release
```

May I understand the rationale behind this statement? It's less obvious than the 1st item where D2D copies always require a system scope release.

Reply:
If you refer the code before my patch, it would only go into the logic to create a new marker ONLY if there exists a depAsyncOp (!= nullptr).
The new logic says that even if there's no depAsyncOn, enqueues a marker with system scope release anyway if the queue's nextSyncNeedsSysRelease flag is true.  This is the case that this change is trying to address.

Reply:
thanks. it makes sense now.

so we could imagine the offending scenario as D2D copy followed by a kernel on the target GPU, right?

Reply:
For D2D, the scenario exposing the problem looks like:

kernel writes to buffer_A (nextSyncNeedsSysRelease set to true) --> copy buffer_A to buffer_B on the same device (nextSyncNeedsSysRelease incorrectly cleared) --> copy buffer_B to host (incorrect since no system release)

In other case, the scenario looks like:

kernel writes to buffer_A (nextSyncNeedsSysRelease set to true) --> user created event/marker clears the async ops tables --> copy buffer_A to host (no async op, the copy logic didn't insert a marker with system release)

Reply:
@scchan I'm wondering would the aggregate effect of the PR causing we _always_ use system scope barriers, thus effectively turning off `HCC_OPT_FLUSH`?

Reply:
no, I don't think so.  They don't interfere with HCC_OPT_FLUSH.
For the D2D case, it's a bug that hcc runtime is not tracking the state of nextSyncNeedsSysRelease correctly after a device memory to device memory copy.  
For the case where the async op table is empty but the queue says that the nextSyncNeedsSysRelease, the guard for creating a new sync is incorrect.